File: PolyAlphabeticCipher/VernamCipher/Decryption/VernamCipher.c
// Improvement of Vernam Cipher
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void removeSpaces(char *str) {
    char *dest = str;
    char *src = str;
    
    while (*src != '\0') {
        if (!isspace(*src)) {
            *dest = *src;
            dest++;
        }
        src++;
    }
    
    *dest = '\0';
}

int main() {
    int num;
    char alphabets[26], cipher_text[100], random_key[100];
    printf("Enter a cipher text: ");
    scanf(" %[^\n]", cipher_text);
    printf("Enter a random key: ");
    scanf(" %[^\n]", random_key);
    removeSpaces(cipher_text);
    removeSpaces(random_key);
    int length = strlen(cipher_text);
    int cipher_text_arr[length];
    int random_key_arr[length];
    int plain_text_arr[length];
    char plain_text[100] = "";
    for (int j = 0; j < 26; j++) {
        alphabets[j] = (char)toupper('A' + j);
    }
    if (strlen(cipher_text) == strlen(random_key)) {
        for (int i = 0; i < strlen(cipher_text); i++) {
            for (int k = 0; k < 26; k++) {
                if (toupper(cipher_text[i]) == alphabets[k]) {
                    cipher_text_arr[i] = k;
                }
            }
        }
        for (int i = 0; i < strlen(random_key); i++) {
            for (int k = 0; k < 26; k++) {
                if (toupper(random_key[i]) == alphabets[k]) {
                    random_key_arr[i] = k;
                }
            }
        }
        for (int i=0; i<length; i++) {
            plain_text_arr[i] = (random_key_arr[i]^cipher_text_arr[i])%26;   
        }
        for (int i = 0; i < length; i++) {
            plain_text[i] = alphabets[plain_text_arr[i]];
            }
        printf("Encrypted Word: %s", plain_text);
    }
    
    return 0;
}

File: PolyAlphabeticCipher/VernamCipher/Decryption/VernamCipher.js
function index_array(array, alpharray) {
  const return_array = [];
  for (let i = 0; i < array.length; i++) {
    for (let j = 0; j < alpharray.length; j++) {
      if (array[i] === alpharray[j]) {
        return_array.push(j);
        break;
      }
    }
  }
  return return_array;
}

function VernamCipher(cipher, random) {
  var result = "";
  const alphabets = [];
  for (let i = 65; i <= 90; i++) {
    alphabets.push(String.fromCharCode(i));
  }
  var cipher_index_array = index_array(cipher, alphabets);
  var random_index_array = index_array(random, alphabets);
  const enc_key_index = [];
  for (let i = 0; i < cipher.length; i++) {
    enc_key_index[i] = cipher_index_array[i] ^ random_index_array[i];
  }
  for (let i = 0; i < cipher.length; i++) {
    result += alphabets[enc_key_index[i]];
  }
  return "Decrypted Word: " + result;
}

function read_input() {
  var cipher_text = prompt("Enter the cipher text: ").toUpperCase().split("");
  var random_key = prompt("Enter the random key: ").toUpperCase().split("");
  if (cipher_text.length === random_key.length) {
    console.log(VernamCipher(cipher_text, random_key));
  } else {
    console.log("Enter a valid cipher text and random key pair");
  }
}

read_input();
